### ISSUE TITLE
Retry

### ISSUE THREAD

**Issue by Jericho** [Pathoschild/FluentHttpClient#22]
It would be convenient to have a way to automatically retry http requests when necessary. The best example I can think of is: some APIs will return a HTTP 429 status code (which means 'TOO MANY REQUESTS') when you exceed the number of request within a certain period of time. The vendor will typically tell you to wait a little bit before retrying you call. Some vendors even include how long you should wait in the response. Another example, is when the API returns a code that indicate a transient problem occurred and you should retry your call.

I propose the following:

1. Add a `IRetryStrategy` interface. This interface is really simple; only two methods: `ShouldRetry` and `GetNextDelay`.
2. Provide a default implementation called `NoRetry` which does not retry. This would match the library's current behavior.
3. Add `SetRetryStrategy(IRetryStrategy retryStrategy)` method on 'IClient' which allows specifying the retry strategy used for all subsequent calls
4. Add `WithRetryStrategy(IRetryStrategy retryStrategy)` method on 'IRequest' which allows specifying the retry strategy used for the current call
5. Developers can create new classes based on `IRetryStrategy` to implement their own retry logic. Their class determines which calls can be retried, the max number of time a call can be retried, how long to pause between each attempt, etc.

I will submit a PR. I can also provide a custom retry strategy I used on a previous project for illustration purposes.

**Comment by Pathoschild** [comment]
@Jericho I'd like to support using [Polly](https://github.com/App-vNext/Polly) and similar libraries, which are pretty powerful. Unfortunately PR #23 wouldn't support that.

Here are a few possible approaches off the top of my head, though I haven't prototyped any of them. Let me know what you think, or if you have different ideas.

## Option 1: multiple response filters
### Proposed implementation
We could maybe create an interface like this:
```cs
/// <summary>A filter which overrides the response received from the server.</summary>
/// <remarks>This is typically used to implement retry policies, error-handling, etc.</remarks>
public interface IHttpResponseFilter
{
    /// <summary>Whether this filter will handle the specified HTTP response.</summary>
    /// <param name="response">The received HTTP response.</param>
    bool CanHandle(HttpResponseMessage response);

    /// <summary>Intercepts an HTTP response after it's received.</summary>
    /// <param name="request">The HTTP request submitted to the server.</param>
    /// <param name="response">The HTTP response received from the server.</param>
    /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
    /// <returns>The modified HTTP response.</returns>
    Task<HttpResponseMessage> Handle(HttpRequestMessage request, HttpResponseMessage response, Func<HttpRequestMessage, Task<HttpResponseMessage>> send);
}
```

...which would let us implement a retry filter that automatically retries after 1, 5, and 10 seconds when it receives an HTTP 429 response. (This could be expanded to make the intervals or even the retry policy configurable, it's just an example.)

```cs
/// <summary>An HTTP response filter which automatically retries when the request is rate-limited with a backoff interval.</summary>
public class RetryWhenRateLimitedFilter : IResponseFilter
{
    /*********
    ** Properties
    *********/
    /// <summary>The time to wait before each retry attempt.</summary>
    private TimeSpan[] RetryIntervals = { TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(5), TimeSpan.FromSeconds(10) };


    /*********
    ** Public methods
    *********/
    /// <summary>Whether this filter will handle the specified HTTP response.</summary>
    /// <param name="response">The received HTTP response.</param>
    public bool CanHandle(HttpResponseMessage response)
    {
        return (int)response.StatusCode == 429;
    }

    /// <summary>Intercepts an HTTP response after it's received.</summary>
    /// <param name="request">The HTTP request submitted to the server.</param>
    /// <param name="response">The HTTP response received from the server.</param>
    /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
    /// <returns>The modified HTTP response.</returns>
    public Task<HttpResponseMessage> Handle(HttpRequestMessage request, HttpResponseMessage response, Func<HttpRequestMessage, Task<HttpResponseMessage>> send)
    {
        return Policy
            .HandleResult<HttpResponseMessage>(this.CanHandle)
            .WaitAndRetry(this.RetryIntervals)
            .ExecuteAsync(async () => await send(request));
    }
}
```

### Benefits
This approach...
* lets you use [Polly](https://github.com/App-vNext/Polly)'s full transient-fault-handling power, another library, or your own logic;
* lets you define multiple retry policies (e.g. handle HTTP 429 and HTTP 500 differently);
* is powerful and easy-to-use for other scenarios like error-handling and response rewriting.

### Caveats
* We'd need some way to coordinate response filters. For example, the `RetryWhenRateLimitedFilter` might result in HTTP 500 which is handled by a different filter which was checked first. One option is to detect when a response filter changed the result, and give the filters another chance to intercept it.
  * Sub-caveat: the solution to the above caveat introduces the possibility of infinite loops (e.g. filter A changes the response in such a way that filter B handles it, which then changes it such that filter A handles it, etc). I don't see an immediate solution that doesn't cause more problems, but I might think of one later.

## Option 2: single retry policy
### Proposed implementation
Option 1 is powerful but introduces new problems. We could extend your original pull request to support a similar interface:

```cs
/// <summary>Defines how the HTTP client should retry requests to handle transient failures and errors.</summary>
public interface IRetryStrategy
{
    /// <summary>Get whether the request should be retried.</summary>
    /// <param name="response">The received HTTP response.</param>
    bool ShouldRetry(HttpResponseMessage response);

    /// <summary>Handle an HTTP request that should be retried.</summary>
    /// <param name="request">The HTTP request submitted to the server.</param>
    /// <param name="response">The HTTP response received from the server.</param>
    /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
    /// <returns>The modified HTTP response.</returns>
    Task<HttpResponseMessage> Handle(HttpRequestMessage request, HttpResponseMessage response, Func<HttpRequestMessage, Task<HttpResponseMessage>> send);
}
```

...which would let us implement the same retry logic as above. (This could be expanded to let you pass in the Polly policy, it's just an example.)
```cs
/// <summary>A retry strategy that applies retry intervals when the server returns HTTP 429.</summary>
public class IntervalRetryStrategy : IRetryStrategy
{
    /*********
    ** Properties
    *********/
    /// <summary>The time to wait before each retry attempt.</summary>
    private TimeSpan[] RetryIntervals = { TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(5), TimeSpan.FromSeconds(10) };


    /*********
    ** Public methods
    *********/
    /// <summary>Get whether the request should be retried.</summary>
    /// <param name="response">The received HTTP response.</param>
    public bool ShouldRetry(HttpResponseMessage response)
    {
        return (int)response.StatusCode == 429;
    }

    /// <summary>Handle an HTTP request that should be retried.</summary>
    /// <param name="request">The HTTP request submitted to the server.</param>
    /// <param name="response">The HTTP response received from the server.</param>
    /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
    /// <returns>The modified HTTP response.</returns>
    public Task<HttpResponseMessage> Handle(HttpRequestMessage request, HttpResponseMessage response, Func<HttpRequestMessage, Task<HttpResponseMessage>> send)
    {
        return Policy
            .HandleResult<HttpResponseMessage>(this.ShouldRetry)
            .WaitAndRetry(this.RetryIntervals)
            .ExecuteAsync(async () => await send(request));
    }
}
```

### Benefits
This approach...
* lets you use [Polly](https://github.com/App-vNext/Polly)'s full transient-fault-handling power, another library, or your own logic;
* doesn't need any coordination logic (like option 1 does), since only one filter is allowed;
* is easy to implement.

### Caveats
* You can't have multiple retry strategies. For example, you'll need your own `IRetryStrategy` if you want to handle two status codes differently; you wouldn't just register two prebuilt filters like `RetryWhenRateLimitedFilter` and `UseCacheWhenServerDownFilter`.
* You can't reuse this for other scenarios like you can with option 1, but that's probably fine.

**Comment by Jericho** [comment]
As you said, option 1 introduces a lot of complexity and requires coordination between strategies. Like you said, what if one strategy want s to handle the failed called performed by another strategy but also, what if more than one strategy want to retry the same call, etc. To be honest, this would be more complex than what I had in mind.

Option 2 is essentially what I implemented in my PR except I didn't use `Polly` and instead I wrote a short snippet of C# code to execute and retry calls (I'm certainly not opposed to use `Polly` if you feel that's an important requirement). By the way, this is the same snippet that has been successfully working in my [library](https://github.com/Jericho/StrongGrid/blob/master/Source/StrongGrid/Client.cs#L545) (ultimately I would like to replace my snippet with the retry strategy in FluentHttpClient). By the way, I used [Microsoft Patern and Practice's retry strategy](https://msdn.microsoft.com/en-us/library/hh680943(v=PandP.50).aspx) as my inspiration.

One thing that was important to me when designing the retry policy was to separate the description of the policy (when should we attempt to retry, the max number of times to retry, how long should we pause between retries, etc.) from the code that actually implements the retry logic. I wanted to avoid the developers having to write their own logic, or even having to use a nuget package such as Polly to implement their own logic. I agree that Polly makes things simple and can be learned quickly, but still it's one more thing for them to deal with. That's why I'm not a fan of asking developers to write their retry logic in `public Task<HttpResponseMessage> Handle(...)`. 

By the way, let me show you an [example of a custom retry strategy](https://github.com/Jericho/StrongGrid/blob/master/Source/StrongGrid/Utilities/SendGridRetryStrategy.cs). This strategy will retry up to 5 times any http call that result in HttStatus 429 and also it calculates the "delay" based on the content of a header called "X-RateLimit-Reset" which is returned by the 3rd party API that my library interacts with. 

I propose improving my PR with the following compromise: how about we use a IRetryStrategy interface similar to the one in my submission (and somewhat similar to the one in your option 2) and also improve the PR I submitted to use `Polly` instead of my custom code snippet.

**Comment by Jericho** [comment]
Here's what I came up with:

1. Add reference to `Polly` nuget package
2. Simplify the IRetryStrategy interface like so:
```CSharp
public interface IRetryStrategy
{
    int MaxRetriesCount { get; }
    bool ShouldRetry(HttpResponseMessage response);
    TimeSpan GetNextDelay(int attempt);
}
```
3. Replace my custom retry logic code with the following `Polly` policy:
```CSharp
var retryPolicy = Policy
    .HandleResult<HttpResponseMessage>(responseMessage => RetryStrategy.ShouldRetry(responseMessage))
    .WaitAndRetryAsync(RetryStrategy.MaxRetriesCount, retryAttempt => RetryStrategy.GetNextDelay(retryAttempt));

var response = await retryPolicy.ExecuteAsync(async () => await request.ConfigureAwait(false));
```

**Comment by Pathoschild** [comment]
I don't want the main package to depend on Polly, I just want to let developers use libraries like it if they want to. :)

I understand not wanting developers to write their own logic unnecessarily. How about this compromise: we can create `IRetryStrategy` per my previous comment, but we'll also have an abstract class that enables the simplified interface for delay strategies from your PR:
```cs
/// <summary>An abstract retry strategy that adds delays between failed attempts.</summary>
public abstract class BaseDelayedRetryStrategy : IRetryStrategy
{
    /*********
    ** Public methods
    *********/
    /// <summary>Get whether the request should be retried.</summary>
    /// <param name="response">The received HTTP response.</param>
    public bool ShouldRetry(HttpResponseMessage response)
    {
        return this.ShouldRetry(1, response);
    }

    /// <summary>Handle an HTTP request that should be retried.</summary>
    /// <param name="request">The HTTP request submitted to the server.</param>
    /// <param name="response">The HTTP response received from the server.</param>
    /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
    /// <returns>The modified HTTP response.</returns>
    public async Task<HttpResponseMessage> Handle(HttpRequestMessage request, HttpResponseMessage response, Func<HttpRequestMessage, Task<HttpResponseMessage>> send)
    {
        var attempt = 1;
        while (attempt == 1 || this.ShouldRetry(attempt, response))
        {
            if (attempt >= 25)
                throw new ApiException(this, response, "Too many attempts");
            var delay = this.GetNextDelay(attempt, response);
            if (delay.TotalMilliseconds > 0)
                await Task.Delay(delay).ConfigureAwait(false);
            response = await send(request);
            attempt++;
        }
    }

    /*********
    ** Protected methods
    *********/
    /// <summary>Get whether the request should be retried.</summary>
    /// <param name="attempt">The number of attempts carried out so far (starting at 1 for the first retry).</param>
    /// <param name="response">The HTTP response received from the server.</param>
    public abstract bool ShouldRetry(int attempt, HttpResponseMessage response);

    /// <summary>Get the delay we should wait until trying again after an unsuccessful attempt.</summary>
    /// <param name="attempt">The number of attempts carried out so far (starting at 1 for the first retry).</param>
    /// <param name="response">The HTTP response received from the server.</param>
    public abstract TimeSpan GetNextDelay(int attempt, HttpResponseMessage response);
}
```

Then the `SendGridRetryStrategy` you linked is only slightly changed (essentially subclass `BaseDelayedRetryStrategy` instead of implementing `IRetryStrategy` directly):
```cs
/// <summary>
/// Implements IRetryStrategy with back off based on a maximum number of retries and
/// a wait time derived from the "X-RateLimit-Reset" responseheader. The value in this
/// header contains the date and time (expressed as the number of seconds since midnight
/// on January 1st 1970) when the next attempt can take place.
/// </summary>
public class SendGridRetryStrategy : BaseDelayedRetryStrategy
{
	/*********
	** Properties
	*********/
	private const int DEFAULT_MAX_RETRIES = 5;
	private const HttpStatusCode TOO_MANY_REQUESTS = (HttpStatusCode)429;
	private static readonly TimeSpan DEFAULT_DELAY = TimeSpan.FromSeconds(1);

	private readonly int MaxAttempts;
	private readonly ISystemClock Clock;


	/*********
	** Public methods
	*********/
	/// <summary>Construct an instance.</summary>
	/// <param name="maxAttempts">The maximum attempts.</param>
	/// <param name="systemClock">The system clock.</param>
	public SendGridRetryStrategy(int maxAttempts = DEFAULT_MAX_RETRIES, ISystemClock systemClock = null)
	{
		this.MaxAttempts = maxAttempts;
		this.Clock = systemClock ?? SystemClock.Instance;
	}

    /// <summary>Get whether the request should be retried.</summary>
    /// <param name="attempt">The number of attempts carried out so far (starting at 1 for the first retry).</param>
    /// <param name="response">The HTTP response received from the server.</param>
    public override bool ShouldRetry(int attempt, HttpResponseMessage response)
    {
		if (attempt >= this.MaxAttempts)
			return false;
		if (response == null)
			return false;
		if (response.StatusCode == TOO_MANY_REQUESTS)
			return true;
		false;
    }

    /// <summary>Get the delay we should wait until trying again after an unsuccessful attempt.</summary>
    /// <param name="attempt">The number of attempts carried out so far (starting at 1 for the first retry).</param>
    /// <param name="response">The HTTP response received from the server.</param>
    public override TimeSpan GetNextDelay(int attempt, HttpResponseMessage response)
    {
		// default value in case the 'reset' time is missing from HTTP headers
		var waitTime = DEFAULT_DELAY;

		// get the 'reset' time from the HTTP headers (if present)
		if (response?.Headers != null)
		{
			var values = response.Headers.Where(h => h.Key == "X-RateLimit-Reset");
			if (values.Any())
			{
				var reset = long.Parse(values.First().Value.First());
				waitTime = reset.FromUnixTime().Subtract(this.Clock.UtcNow);
			}
		}

		// make sure the wait time is valid
		if (waitTime.TotalMilliseconds < 0)
			waitTime = DEFAULT_DELAY;

		// totally arbitrary: make sure we don't wait more than a 'reasonable' amount of time
		if (waitTime.TotalSeconds > 5)
			waitTime = TimeSpan.FromSeconds(5);

		return waitTime;
    }
}
```

**Comment by Jericho** [comment]
What is the benefit to the developer if we offer them the possibility of using a "retry library" of their choice? To be honest, I don't think they should care **how** the retry logic is implemented in the FluentHttpClient library. In other words, they shouldn't care whether we write custom retry logic or we use the `Polly` nuget package. The only thing they should care about is that HTTP requests are automatically retried when the response meets criteria that they have defined (`HttpStatusCode == 429` for example).

I really don't see a benefit but maybe you have a use case in mind that I am not thinking of?

I have submitted a PR to demonstrate what it would look like if we took a dependency to the `Polly` nuget package.

**Comment by Pathoschild** [comment]
Delayed retry is a common strategy, but it's not the only one. For example:

* **Fallback strategy 1:** if the server is overloaded or offline, call an alternate server instead.

* **Fallback strategy 2:** if the server is overloaded or offline, return a cached response.

* **Circuit breaker strategy:** stop sending requests to the server for a specified interval — because if this request failed with a rate limit, so will the next one. And if it failed because the server is overloaded, a delayed retry strategy can cause cascading failures. During the interval, it might use a fallback strategy.

* **Custom strategies:** one of my upcoming experiments involves a proxy server which fetches project releases from the GitHub API, caches it for use by an update checker (since there will be too many requests to hit the GitHub API every time), and returns a simplified result.

  When GitHub returns a rate limit error, the retry strategy I'm considering would look like this:

  1. If the response headers indicate the limit will roll over soon, delay and retry.
  2. Otherwise if we're under the rate limit as an anonymous user, remove the auth key and retry.
  3. Otherwise if we have a recent cached response, return that.
  3. Otherwise redirect the request to the GitHub API, so it counts against their anonymous IP.

  This might use a version of the circuit breaker — since we know when the rate limits will roll over, we can sometimes skip straight to one of the fallbacks.

These strategies are easy to implement with the `IRetryStrategy` interface (except the circuit breaker which would requires some minor changes), but none of them are possible with the interface from your PR which only supports delayed retries.

So my goal is to make delayed retry simple (as the most common case), while supporting different strategies when needed.

**Comment by Jericho** [comment]
Ok, makes sense. 

How about we have two interfaces:

- **IRetryStrategy** which allows describing the desired retry strategy (max number of retries, delay between retries, etc)
- **IRequestExecutor** (feel free to suggest a different name) which would be responsible for executing HTTP requests with or without retries. This interface would be simple, probably a single method like `public async Task<HttpResponseMessage> ExecuteAsync(Task<HttpResponseMessage> request, IRetryStrategy retryStrategy)` or something similar.

We could provide the following implementations of the IRequestExecutor interface:
- **ExecutorWithRetries** which would use the retry loop from my PR and the RetryStrategy provided by developer
- **ExecutorNoRetries** which would simply execute the HTTP request and not attempt to retry (matches FluentHttpClient current behavior)
- We could even provide **ExecutorWithPollyRetries** just for the fun of it (but I agree with you, I would prefer not to introduce a dependency to another package)
- Developers could write their own **ExecutorWithFallback**, **ExecutorWithCircuitBreaker**, etc.

This way, we clearly separate the description of the strategy and the execution of the HTTP request. What do you think?

**Comment by Pathoschild** [comment]
I like it. With support for the circuit breaker strategy, the executor might look like this:
```cs
/// <summary>Defines how the HTTP client should dispatch requests and process responses at a low level, for example to handle transient failures and errors. Only one may be used by the client.</summary>
public interface IRequestStrategy
{
    /// <summary>Dispatch an HTTP request.</summary>
    /// <param name="request">The HTTP request to submit to the server.</param>
    /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
    /// <returns>The final HTTP response.</returns>
    Task<HttpResponseMessage> OnRequesting(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> send);
}
```

If I understand what you're suggesting correctly, here's a possible implementation of a configurable retry strategy:
```cs
/// <summary>Configures a request retry strategy.</summary>
public interface IRetryConfig
{
    /// <summary>The maximum number of times to retry a request before failing.</summary>
    int MaxRetries { get; }

    /// <summary>Get whether to retry a request.</summary>
    /// <param name="response">The HTTP response received from the server.</param>
    bool ShouldRetry(HttpResponseMessage response);

    /// <summary>Get the delay before the next retry.</summary>
    /// <param name="retry">The number of the next retry (where the first one is 1).</param>
    /// <param name="response">The HTTP response received from the server.</param>
    TimeSpan GetDelay(int retry, HttpResponseMessage response);
}
```
```cs
/// <summary>A request strategy which retries failed requests with a delay between each attempt.</summary>
public class RetryStrategy : IRequestStrategy
{
    /*********
    ** Properties
    *********/
    /// <summary>The retry configuration.</summary>
    private readonly IRetryConfig Config;


    /*********
    ** Public methods
    *********/
    /// <summary>Construct a default instance.</summary>
    public RetryStrategy()
        // TODO: probably have a list of default status codes, but we'll just check 429 for this example
        : this(response => (int)response.StatusCode == 429, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(5)) {}

    /// <summary>Construct an instance.</summary>
    /// <param name="lambda">A lambda which returns whether a request should be retried.</param>
    /// <param name="intervals">The intervals between each retry attempt.</param>
    public RetryStrategy(Func<HttpResponseMessage> lambda, params TimeSpan[] intervals)
        : this(new RetryIntervalConfig(lambda, intervals)) {}

    /// <summary>Construct an instance.</summary>
    /// <param name="config">The retry configuration.</param>
    public RetryStrategy(IRetryConfig config)
    {
        this.Config = config;
    }

    /// <summary>Dispatch an HTTP request.</summary>
    /// <param name="request">The HTTP request to submit to the server.</param>
    /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
    /// <returns>The final HTTP response.</returns>
    public async Task<HttpResponseMessage> OnRequesting(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> send)
    {
        HttpResponseMessage response = await send(request);
        for(int retry = 1; retry <= this.Config.MaxRetries && this.Config.ShouldRetry(response); retry++)
        {
            TimeSpan delay = this.Config.GetDelay(attempt, response);
            if (delay.TotalMilliseconds > 0)
                await Task.Delay(delay).ConfigureAwait(false);
            response = await send(request);
        }
        return response;
    }
}
```

Does that match what you're thinking?

**Comment by Jericho** [comment]
I feel we are getting close!

Let me show you a few examples of what I have in mind:

1. Match the current fluent client behavior (no retries):
```CSharp
var client = new FluentClient(baseUri, httpClient)
    .SetRequestExecutor(new RequestExecutorWithoutRetries());
```
**Note**: In actuality, the `NoRetries` executor would be the default executor, so you wouldn't need to explicitly set it.

2. Configure fluent client to retry with a custom strategy:
```CSharp
var client = new FluentClient(baseUri, httpClient)
    .SetRequestExecutor(new RequestExecutorWithRetries())
    .SetRetryStrategy(new MyCustomStrategy());
```
3. Override the executor for one specific request (there would be a similar `WithRetryStrategy` to override the strategy):
```CSharp
var response = client
    .GetAsync("http://api.vendor.com/some_resource")
    .WithRequestExecutor(new RequestExecutorWithoutRetries())
    .AsString();
```

By the way, you call my `IRetryStrategy` interface `IRetryConfig`. I think I like your naming better.

Here's my [first draft of what I have in mind](https://github.com/Jericho/FluentHttpClient/commit/5bad7c57c8f24a3c72b5b9c29bbeccc4316d8e56)

**Comment by Pathoschild** [comment]
I like it overall. Just a few quibbles:

* We shouldn't have a separate `SetRetryConfig` method if it only affects one implementation. For example, it would do nothing in this case:
  ```cs
   var client = new FluentClient(baseUri, httpClient)
      .SetRequestExecutor(new CircuitBreakerExecutor())
      .SetRetryConfig(new MyRetryConfig());
  ```
  I would rather add it to the implementation that uses it, which also makes other signatures discoverable:
  ```cs
   // use default retry config
   var client = new FluentClient(baseUri, httpClient)
      .SetRequestExecutor(new RetryExecutor());

   // use custom retry config
   var client = new FluentClient(baseUri, httpClient)
      .SetRequestExecutor(new RetryExecutor(new MyRetryConfig()));

   // use static config (no need to create a custom class)
   var client = new FluentClient(baseUri, httpClient)
      .SetRequestExecutor(new RetryExecutor(
         shouldRetry: r => response => (int)response.StatusCode == 429,
         intervals: new[] { TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(2), TimeSpan.FromSeconds(5) }
      ));
  ```
* I'm not sure about the 'executor' name. The request is executed by the underlying HTTP client; the 'request executor' is only coordinating. I'm also not sure what I'd use instead, though. Request strategy, manager, coordinator, filter? Maybe executor is okay.
* I'd probably tweak the naming a bit (e.g. `RetryExecutor` instead of `RequestExecutorWithRetries`), but we can discuss that last. ;)

**Comment by Jericho** [comment]
**Your quibble 1**: I agree with you that it makes sense for the IRetryConfig to be a property of the 'executor' object rather than a separate `SetRetryConfig` method. However, I think one common scenario will be for developers to specify the IRetryConfig and assume the default executor is used (that's the way I personally expect to use the retry feature since I don't care about CircuitBreaker and other executors).
```CSharp
// What I mean is something along these lines:
var client = new FluentClient(baseUri, httpClient)
    .SetRetryStrategy(new MyRetryConfig());
// Which would be equivalent to:
 var client = new FluentClient(baseUri, httpClient)
    .SetRequestExecutor(new RetryExecutor(new MyRetryConfig()));
```
I just want to provide a simplified way to provide the retry config without worrying about specifying the executor. Makes sense? Maybe an overloaded `SetRequestExecutor` that accepts `iRetryConfig` like this: 
```CSharp
public IClient SetRequestExecutor(this IClient client, IRetryConfig retryConfig)
{
    return client.SetRequestExecutor(new RetryExecutor(retryConfig));
}
```

**You quibble 2**: 'Executor' is just the first name that came to my mind. I like 'RequestCoordinator' also, but I don't have a strong preference.

**Your quibble 3**: I don't have strong preference. How about: `RetryCoordinator`, `NoRetryCoordinator`, `CircuitBreakerCoordinator`, etc.

**My quibble 1**: you gave an example of `RetryExecutor` that accepts a `Func` for shouldRetry and an array for `intervals`. The array is fine for cases when you hard code the delay TimeSpans but I would prefer a `Func` for situations where the delay is not known ahead of time an must be calculated based on information in the response (for example my `SendGridRetryStrategy` calculates the TimeSpan based on the value in a response header).

**Comment by Pathoschild** [comment]
Having an overload for `IRetryConfig` makes sense to me, I'm okay with 'coordinator' (or at least I can't think of anything better so far), and I'm fine having a constructor that takes a `Func` for the delay. I think we're agreed on the main implementation details, then. 🎊 

So to recap based on my understanding:

1. Create `IRequestCoordinator`, which dispatches requests and processes responses at a low level to handle transient failures and errors. It can intercept both outgoing requests and incoming responses. You can only have one coordinator at once.

   ```cs
   /// <summary>Defines how the HTTP client should dispatch requests and process responses at a low level, for example to handle transient failures and errors. Only one may be used by the client.</summary>
    public interface IRequestCoordinator
   {
       /// <summary>Dispatch an HTTP request.</summary>
       /// <param name="request">The HTTP request to submit to the server.</param>
       /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
       /// <returns>The final HTTP response.</returns>
       Task<HttpResponseMessage> OnRequesting(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> send);
   }
   ```

2. Create `RetryCoordinator` (with `IRetryConfig` to configure it), a bundled coordinator which retries matching requests until they work or a retry limit is reached.
```cs
   /// <summary>A request coordinator which retries failed requests with a delay between each attempt.</summary>
   public class RetryCoordinator : IRequestCoordinator
   {
       /*********
       ** Properties
       *********/
       /// <summary>The retry configuration.</summary>
       private readonly IRetryConfig Config;
   
   
       /*********
       ** Public methods
       *********/
       /// <summary>Construct an instance.</summary>
       /// <param name="shouldRetry">A lambda which returns whether a request should be retried.</param>
       /// <param name="intervals">The intervals between each retry attempt.</param>
       public RetryStrategy(Func<HttpResponseMessage> shouldRetry, params TimeSpan[] intervals)
           : this(new RetryConfig(shouldRetry, intervals)) {}

       /// <summary>Construct an instance.</summary>
       /// <param name="shouldRetry">A lambda which returns whether a request should be retried.</param>
       /// <param name="getDelay">Get the time to wait until the next retry.</param>
       public RetryStrategy(Func<HttpResponseMessage> shouldRetry, Func<int, HttpResponseMessage, TimeSpan> getDelay)
           : this(new RetryConfig(shouldRetry, getDelay)) {}
   
       /// <summary>Construct an instance.</summary>
       /// <param name="config">The retry configuration.</param>
       public RetryStrategy(IRetryConfig config)
       {
           this.Config = config;
       }
   
       /// <summary>Dispatch an HTTP request.</summary>   
       /// <param name="request">The HTTP request to submit to the server.</param>   
       /// <param name="send">A lambda which submits an HTTP request and retrieves the response.</param>
       /// <returns>The final HTTP response.</returns>
       public async Task<HttpResponseMessage> OnRequesting(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> send)
       {
           HttpResponseMessage response = await send(request);
           for(int retry = 1; retry <= this.Config.MaxRetries && this.Config.ShouldRetry(response); retry++)
           {
               TimeSpan delay = this.Config.GetDelay(retry, response);
               if (delay.TotalMilliseconds > 0)
                   await Task.Delay(delay).ConfigureAwait(false);
               response = await send(request);
           }
           return response;
       }
   }
```
```cs
/// <summary>Configures a request retry strategy.</summary>
public interface IRetryConfig
{
    /// <summary>The maximum number of times to retry a request before failing.</summary>
    int MaxRetries { get; }

    /// <summary>Get whether to retry a request.</summary>
    /// <param name="retry">The number of the next retry (where the first one is 1).</param>
    /// <param name="response">The HTTP response received from the server.</param>
    bool ShouldRetry(int retry, HttpResponseMessage response);

    /// <summary>Get the delay before the next retry.</summary>
    /// <param name="retry">The number of the next retry (where the first one is 1).</param>
    /// <param name="response">The HTTP response received from the server.</param>
    TimeSpan GetDelay(int retry, HttpResponseMessage response);
}
```

3. Add fluent methods to configure the client's request coordinator:
   ```cs
   // main method
   var client = new FluentClient(…)
      .SetRequestCoordinator(new RetryCoordinator(…));

   // convenience shortcut
   var client = new FluentClient(…)
      .SetRequestCoordinator(new MyRetryConfig(…));
   ```

4. Add equivalent `WithRequestCoordinator` methods for a single request. (Do we need this? Seems a bit overkill.)

5. If no retry coordinator is registered, use a coordinator that just dispatches requests and returns responses as-is (`SimpleCoordinator`?).

Does that sum it up?

**Comment by Jericho** [comment]
Sums it up (almost!) perfectly ;)

One very minor detail: I think the name of the method on the IRequestCoordinator should be a verb like "Execute" for example rather than 'OnRequesting' which to me invokes an event being raised. 

I will be offline over the weekend, but I'll be back on Monday to get started.

**Comment by Pathoschild** [comment]
I'm fine with _Execute_. :)

**Comment by Jericho** [comment]
Do you plan to look at the PR to move this forward?

**Comment by Pathoschild** [comment]
Yep. I rotate between side projects, so my full attention will be back on the PRs in a few days. Let me know if it's urgent (e.g. you need this for an upcoming project), and I'll shuffle my queue. :)

**Comment by Jericho** [comment]
It's not really _urgent_ but I did put my own project on the backburner in order to focus on this so I'm kind of antsy to move forward. Also, I wrote a long comment on the PR to explain the reasons why I am proposing some changes to the FluentHttpClient that would solve problems related to the retry logic. I hope I clearly explained the challenges I faced and that I made a good case for each proposed change and I would like to know if you agree.
